### ISSUE TITLE
Audit example config for UX improvements

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#1965]
A user of Vector [shared their configuration with us](https://github.com/timberio/vector/issues/1721#issuecomment-593104785). I would like to audit their configuration to see if there are ways we can improve the UX. Things I'm looking for:

1. They're using Lua a few times. What upstream options or other transforms would help avoid this?
2. Are there any default behavior changes that would reduce this config?
3. Would #1926 reduce the config size, improve readability, and offer more flexibility?

**Comment by Jeffail** [comment]
First pass, this is a simplified version that should be possible with v0.8.0, where we use swimlanes instead of `field_filter`:

```toml
data_dir = "/var/lib/vector"

[sources.in]
  type = "syslog"
  mode = "tcp"
  address = "0.0.0.0:5240"

[transforms.app_type]
  type = "swimlanes"
  inputs = ["in"]

  [transforms.app_type.lanes.company_api]
    "appname.eq" = "company-api"

  [transforms.app_type.lanes.company_admin]
    "appname.eq" = "company-admin"

  [transforms.app_type.lanes.company_media_proxy]
    "appname.eq" = "company-media-proxy"

  [transforms.app_type.lanes.company_unfurler]
    "appname.eq" = "company-unfurler"

  [transforms.app_type.lanes.audit]
    "appname.eq" = "audit"

[transforms.company_api_json]
type = "json_parser"
inputs = ["app_type.company_api"]
drop_invalid = true

[transforms.company_api_lua]
type = "lua"
inputs = ["company_api_json"]
source = """
event["time"] = math.floor(event["timestamp"])
event["hostname"] = event["host"]
event["metadata_trace_id"] = event["metadata.trace_id"]
event["metadata_guild_id"] = event["metadata.guild_id"]
event["metadata_channel_id"] = event["metadata.channel_id"]
event["metadata_method"] = event["metadata.method"]
"""

[transforms.company_api_clean]
type = "remove_fields"
inputs = ["company_api_lua"]
fields = [
    # company-api
    # ASSUMPTION: We're removing all fields within metadata.
    "metadata",
    # rsyslog
    "timestamp",
    "pid",
    "appname",
    "facility",
    "host",
    "msgid",
    "procid",
    "severity",
    "version"
]

[sinks.company_api_clickhouse]
type = "clickhouse"
inputs = ["company_api_clean"]
host = "http://clickhouse-url:4213"
table = "company_api"

[sinks.company_api_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.company_api_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"

[sinks.company_api_clickhouse.batch]
max_size = 5245000
timeout_secs = 5

[transforms.company_admin_json]
type = "json_parser"
inputs = ["app_type.company_admin"]
drop_invalid = true

[transforms.company_admin_lua]
type = "lua"
inputs = ["company_admin_json"]
source = """
event["time"] = math.floor(event["timestamp"])
event["hostname"] = event["host"]
event["metadata_trace_id"] = event["metadata.trace_id"]
event["metadata_method"] = event["metadata.method"]
"""

[transforms.company_admin_clean]
type = "remove_fields"
inputs = ["company_admin_lua"]
fields = [
    # company-admin
    # ASSUMPTION: We're removing all fields within metadata.
    "metadata",
    # rsyslog
    "timestamp",
    "pid",
    "appname",
    "facility",
    "host",
    "msgid",
    "procid",
    "severity",
    "version"
]

[sinks.company_admin_clickhouse]
type = "clickhouse"
inputs = ["company_admin_clean"]
host = "http://clickhouse-url:4213"
table = "company_admin"

[transforms.company_media_proxy_json]
type = "json_parser"
inputs = ["app_type.company_media_proxy"]
drop_invalid = true

[transforms.company_media_proxy_lua]
type = "lua"
inputs = ["company_media_proxy_json"]
source = """
event["time"] = math.floor(event["ts"])
event["hostname"] = event["host"]
"""

[transforms.company_media_proxy_clean]
type = "remove_fields"
inputs = ["company_media_proxy_lua"]
fields = [
    # company-media-proxy
    "ts",
    "level",
    "path",
    "error",
    "type",
    "height_px",
    "width_px",
    "width",
    "height",
    "ext",
    "size",
    "size_bytes",
    # rsyslog
    "timestamp",
    "pid",
    "appname",
    "facility",
    "host",
    "msgid",
    "procid",
    "severity",
    "version"
]

[sinks.company_media_proxy_clickhouse]
type = "clickhouse"
inputs = ["company_media_proxy_clean"]
host = "http://clickhouse-url:4213"
table = "company_media_proxy"

[sinks.company_media_proxy_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.company_media_proxy_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"

[sinks.company_media_proxy_clickhouse.batch]
max_size = 5245000
timeout_secs = 5

[transforms.company_unfurler_hostname]
type = "lua"
inputs = ["app_type.company_unfurler"]
source = """
event["hostname"] = event["host"]
"""

[transforms.company_unfurler_json]
type = "json_parser"
inputs = ["company_unfurler_hostname"]
drop_invalid = true

[transforms.company_unfurler_filter]
type = "field_filter"
inputs = ["company_unfurler_json"]
field = "msg"
value = "unfurl"

[transforms.company_unfurler_lua]
type = "lua"
inputs = ["company_unfurler_filter"]
source = """
event["time"] = event["ts"]:gsub("%+00:00", ""):gsub("%.%d+$", ""):gsub("T", " ")
"""

[transforms.company_unfurler_clean]
type = "remove_fields"
inputs = ["company_unfurler_lua"]
fields = [
  # company-unfurler
  "ts",
  "level",
  "duration",
  # rsyslog
  "timestamp",
  "pid",
  "appname",
  "facility",
  "msgid",
  "procid",
  "severity",
  "version"
]

[sinks.company_unfurler_clickhouse]
type = "clickhouse"
inputs = ["company_unfurler_clean"]
host = "http://clickhouse-url:4213"
table = "company_unfurler"

[sinks.company_unfurler_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.company_unfurler_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"

[transforms.audit_lua]
type = "lua"
inputs = ["app_type.audit"]
source = """
event["tag"] = event["appname"]
event["hostname"] = event["host"]
event["pid"] = event["pid"]
event["content"] = event["message"]
event["time"] = event["timestamp"]:gsub("Z", ""):gsub("%.%d+$", ""):gsub("T", " ")
"""

[transforms.audit_clean]
type = "remove_fields"
inputs = ["audit_lua"]
fields = [
    # rsyslog
    "timestamp",
    "message",
    "pid",
    "appname",
    "facility",
    "host",
    "msgid",
    "procid",
    "severity",
    "version"
]

[sinks.audit_clickhouse]
type = "clickhouse"
inputs = ["audit_clean"]
host = "http://clickhouse-url:4213"
table = "syslog"

[sinks.audit_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.audit_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"
```

Still lots of lines, primarily the excessive deletes, although we can save a few lines by deleting `metadata` thanks to the new nested path field work.

**Comment by Jeffail** [comment]
Second pass, this would require a new condition type (a few minutes to implement), the new encoding fields from https://github.com/timberio/vector/issues/1448 to drop `appname` at the sink level, and a new mapping language:

```toml
data_dir = "/var/lib/vector"

[sources.in]
  type = "syslog"
  mode = "tcp"
  address = "0.0.0.0:5240"

[transforms.app_type]
  type = "swimlanes"
  inputs = ["in"]

  [transforms.app_type.lanes.company]
    "appname.prefix" = "company-"

  [transforms.app_type.lanes.audit]
    "appname.eq" = "audit"

[transforms.json_company]
  type = "json_parser"
  inputs = ["app_type.company"]
  drop_invalid = true

[transforms.mapped_app_type]
  type = "idml_mapper"
  inputs = ["json_company", "app_type.audit"]
  map = "/etc/vector/app_mapping.idml"

[transforms.mapped_app_type]
  type = "swimlanes"
  inputs = ["mapped_app_type"]

  [transforms.mapped_app_type.lanes.company_api]
    "appname.eq" = "company-api"

  [transforms.mapped_app_type.lanes.company_admin]
    "appname.eq" = "company-admin"

  [transforms.mapped_app_type.lanes.company_media_proxy]
    "appname.eq" = "company-media-proxy"

  [transforms.mapped_app_type.lanes.company_unfurler]
    "appname.eq" = "company-unfurler"
    "msg.eq" = "unfurl"

  [transforms.app_type.lanes.audit]
    "appname.eq" = "audit"

# Company API
#-------------------------------------------------------------------------------
[sinks.company_api_clickhouse]
type = "clickhouse"
inputs = ["company_api_clean"]
host = "http://clickhouse-url:4213"
table = "company_api"

[sinks.company_api_clickhouse.encoding]
except_fields = [ "appname" ]

[sinks.company_api_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.company_api_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"

[sinks.company_api_clickhouse.batch]
max_size = 5245000
timeout_secs = 5

# Company Admin
#-------------------------------------------------------------------------------
[sinks.company_admin_clickhouse]
type = "clickhouse"
inputs = ["company_admin_clean"]
host = "http://clickhouse-url:4213"
table = "company_admin"

[sinks.company_admin_clickhouse.encoding]
except_fields = [ "appname" ]

# Company Media Proxy
#-------------------------------------------------------------------------------
[sinks.company_media_proxy_clickhouse]
type = "clickhouse"
inputs = ["company_media_proxy_clean"]
host = "http://clickhouse-url:4213"
table = "company_media_proxy"

[sinks.company_media_proxy_clickhouse.encoding]
except_fields = [ "appname" ]

[sinks.company_media_proxy_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.company_media_proxy_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"

[sinks.company_media_proxy_clickhouse.batch]
max_size = 5245000
timeout_secs = 5

# Company Unfurler
#-------------------------------------------------------------------------------
[sinks.company_unfurler_clickhouse]
type = "clickhouse"
inputs = ["company_unfurler_clean"]
host = "http://clickhouse-url:4213"
table = "company_unfurler"

[sinks.company_unfurler_clickhouse.encoding]
except_fields = [ "appname" ]

[sinks.company_unfurler_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.company_unfurler_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"

# Audit
#-------------------------------------------------------------------------------
[sinks.audit_clickhouse]
type = "clickhouse"
inputs = ["audit_mapped"]
host = "http://clickhouse-url:4213"
table = "syslog"

[sinks.audit_clickhouse.encoding]
except_fields = [ "appname" ]

[sinks.audit_clickhouse.request]
in_flight_limit = 100
rate_limit_duration_secs = 5
rate_limit_num = 60

[sinks.audit_clickhouse.buffer]
type = "memory"
max_events = 1000
when_full = "block"
```

And the map might look something like this:

```
[rsyslog]

timestamp = deleted()
pid = deleted()
facility = deleted()
msgid = deleted()
procid = deleted()
severity = deleted()
version = deleted()

[company-api]

time = math.floor(timestamp)

hostname = host
host = deleted()

metadata_trace_id = metadata.trace_id
metadata_guild_id = metadata.guild_id
metadata_channel_id = metadata.channel_id
metadata_method = metadata.method
metadata = deleted()

root = root.apply("rsyslog")

[company-admin]

time = math.floor(timestamp)

hostname = host
host = deleted()

metadata_trace_id = metadata.trace_id
metadata_method = metadata.method
metadata = deleted()

[company-media-proxy]

time = math.floor(ts)
ts = deleted()

level = deleted()

hostname = host
host = deleted()

path = deleted()
error = deleted()
type = deleted()
height_px = deleted()
width_px = deleted()
width = deleted()
height = deleted()
ext = deleted()
size = deleted()
size_bytes = deleted()

[company-unfurler]

time = math.floor(ts)
ts = deleted()

level = deleted()
duration = deleted()

[audit]

tag = appname
pid = pid.string()
time = timestamp:gsub("Z", ""):gsub("%.%d+$", ""):gsub("T", " ")
content = message
message = deleted()

hostname = host
host = deleted()

[main]

# Copy everything over from the input. This might be much more complicated than
# instead moving over from a whitelist of fields. However, I'd need more context
# around the use case to make a call on that.
root = this

root = match appname
    this == company-api => root.apply("company-api")
    this == company-admin => root.apply("company-admin")
    this == company-media-proxy => root.apply("company-media-proxy")
    this == company-unfurler => root.apply("company-unfurler")
    this == audit => root.apply("audit")

root = root.apply("rsyslog")
```

This saves a lot of lines as it allows users to define certain maps as a function and reuse them, which isn't possible with transforms themselves. It also combines the behaviors from both the `lua` and `remove_fields` transforms into one map, which helps a lot with readability and maintainability.

I've had to make a few assumptions here and the mapping it predominantly deleting fields. However, I suspect for this particular use case the map could be even simpler if we instead focused on copying over whitelisted fields rather than copying everything and doesn't exhaustive deletes.

**Comment by Jeffail** [comment]
Also it's worth noting that the mapping shown here is also possible with the `lua` transform. Given they're already using lua, if the goal were to reduce the size and complexity of the config, then I would recommend using this second example and moving all the mapping logic into lua. The reason why the original config can't be dealt with this way is because it requires a diamond shape of transforms that feeds the map:

```
swimlanes --> company-* -> json_parser --> mapping_logic -> swimlanes ->
          \-> audit -------------------/
```

Which is very awkward to do with `field_filter` but pretty easy with `swimlanes`.

This brings us back to "why implement a mapping language?" Where we go back to performance and determinism, which helps with maintainability.

**Comment by binarylogic** [comment]
Thanks! I think this is a great exercise. I actually found the second example harder to follow than both the original and first example you provided 😄 . What made it hard to follow is:

1. Intermingling all of the streams. For me, it's easier to say think in the context of individual streams because their data model is different. I can put on the mental blinders and think in terms of one data model and one stream.
2. The use of 2 separate `swimlanes` transforms. It took more effort than I'd like to figure out that the `company` streams needed to be JSON decoded.

In general though. I like the idea of:

1. sources -> one big script -> sinks
2. Or, sources -> split (swimlanes) -> TOML file for each stream

If we were to recommend 2, then having a single script that _fully_ transforms the data (add/remove/rename) would make all of this easier. And for that, I agree that your mapping language is much easier to follow.

**Comment by Jeffail** [comment]
Yeah I agree, example 2 leans into reducing lines more than readability (outside of the mapping part). Since we're now able to break the configs down it makes more sense to go in that direction and have an isolated config file for each `appname` type.

Another thing that we could do to make it even simpler is add a json parser function to the mapping language itself. That also avoids the nastiness of parsing the `message` field and potentially overriding fields before you have a chance to map them.

**Comment by binarylogic** [comment]
Yep. This is why I'm leaning more and more towards a syntax like that. I still think there is a place for ` json_parser` transform, especially for new users, but this simple syntax is so much nicer for complex use cases.

**Comment by Hoverbear** [comment]
@Jeffail is this still on your plate?
